QHyper/converter.py

```diff
@@ -147,60 +151,41 @@ def to_dimod_qubo(problem: Problem, lagrange_multiplier: float = 10
                     bqm.to_qubo())  # (qubo, offset)
 
-    # TODO: Refactor is needed
-    @staticmethod
-    def to_dqm(problem: Problem) -> DiscreteQuadraticModel:
-        if hasattr(problem, "one_hot_encoding"):
-            if problem.one_hot_encoding:
-                return Converter._dqm_from_one_hot_representation(problem)
-            return Converter._dqm_from_discrete_representation(problem)
-        return Converter._dqm_from_one_hot_representation(problem)
 
     @staticmethod
-    def _dqm_from_discrete_representation(problem: Problem) -> DiscreteQuadraticModel:
-        dqm = dimod.DiscreteQuadraticModel()
+    def to_dqm(problem: Problem, cases: int = 1) -> DiscreteQuadraticModel:
 
-        for var in problem.variables:
-            var = str(var)
-            if var not in dqm.variables:
-                dqm.add_variable(problem.cases, var)
+        def binary_to_discrete(v: str) -> str:
+            id = int(v[1:])
+            discrete_id = id // cases
+            return f"x{discrete_id}"
 
-        for vars, bias in problem.objective_function.as_dict().items():
-            x_i, x_j = vars
-            xi_idx: int = cast(int, dqm.variables.index(x_i))
-            xj_idx: int = cast(int, dqm.variables.index(x_j))
+        def extract_number(element) -> int:
+            match = re.search(r'(\d+)', element)
+            return int(match.group(1))
 
-            # We're skipping the linear terms
-            if xi_idx == xj_idx:
-                continue
-
-            dqm.set_quadratic(
-                dqm.variables[xi_idx],
-                dqm.variables[xj_idx],
-                {(case, case): bias for case in range(problem.cases)},
+        if problem.constraints:
+            warnings.warn(
+                "Defined problem has constraints. DQM does not support constraints, it only supports objective functions!",
+                ProblemWarning
             )
 
-        return dqm
-
-    @staticmethod
-    def _dqm_from_one_hot_representation(problem: Problem) -> DiscreteQuadraticModel:
         dqm = dimod.DiscreteQuadraticModel()
+        objective_function_variables = sorted(problem.objective_function.get_variables(),
+                     key=extract_number)
 
-        CASES_OFFSET = 1 if problem.cases == 1 else 0
+        variables = [
+            binary_to_discrete(str(v)) for v in objective_function_variables[:: cases]
+            ]
 
-        def binary_to_discrete(v: str) -> str:
-            id = int(v[1:])
-            discrete_id = id // problem.cases
-            return f"x{discrete_id}"
+        cases_offset = 1 if cases == 1 else 0
 
-        variables_discrete = [
-            binary_to_discrete(str(v)) for v in problem.variables[:: problem.cases]
-        ]
-        for var in variables_discrete:
-            if var not in dqm.variables:
-                dqm.add_variable(problem.cases + CASES_OFFSET, var)
+        for variable in variables:
+            variable = str(variable)
```

Comment:
variables are strings right?

Reply:
Sometimes it can be passed as a string and sometimes as a sympy symbol, so there is conversion to string

Reply:
variables are created from binary_to_discrete which returns only str right?